File: leetcode/124.py
class Solution(object):
    def maxPathSum(self, root):
        ans = [-3e9]
        def f(root):
            if root == None:    return 0
            l, r = max(0, f(root.left)), max(0, f(root.right))
            ans[0] = max(ans[0], l + r + root.val)
            return max(l, r) + root.val
        f(root)
        return ans[0]

File: leetcode/124.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    int ans;
public:
    int dfs(TreeNode* rt) {
        if (!rt)    return 0;
        int t1 = 0, t2 = 0;
        t1 = max(t1, dfs(rt -> left));
        t2 = max(t2, dfs(rt -> right));
        ans = max(ans, t1 + t2 + rt -> val);
        return max(t1, t2) + rt -> val;
    }
    int maxPathSum(TreeNode* root) {
        ans = -1e9;
        dfs(root);
        return ans;
    }
};